File: day2.php
<?php
declare(strict_types = 1);

/**
 * Possible symbols
 */
enum Symbol
{
    case Rock;
    case Paper;
    case Scissors;
}

/**
 * Scores for playing a symbol
 */
enum SymbolScore: int
{
    case Rock = 1;
    case Paper = 2;
    case Scissors = 3;
}

/**
 * Match scores
 */
enum MatchScore: int
{
    case Loss = 0;
    case Draw = 3;
    case Win = 6;
}


/**
 * scoreRound : Determine the player score for a round. 
 *     Rock > Scissors, Scissors > Paper, Paper > Rock
 *     Player scores Rock = 1, Paper = 2, Scissors = 3
 *     Player scores Loss = 0, Draw = 3, Win = 6
 * @param Symbol $elf       Elf symbol
 * @param Symbol $player    Player symbol
 * 
 * @return int
 */
function scoreRound( Symbol $elf, Symbol $player): int
{
    $score = 0;
    switch ($player) {
        case Symbol::Rock:
            // Add score for playing symbol
            $score += SymbolScore::Rock->value;
            // Add score for win, draw or loss
            if ($elf == Symbol::Rock) {
                $score += MatchScore::Draw->value; 
            } else {
                $score += ($elf == Symbol::Scissors) ? MatchScore::Win->value : MatchScore::Loss->value;
            } 
            break;
        case Symbol::Paper:
            // Add score for playing symbol
            $score += SymbolScore::Paper->value;
            // Add score for win, draw or loss
            if ($elf == Symbol::Paper) {
                $score += MatchScore::Draw->value; 
            } else {
                $score += ($elf == Symbol::Rock) ? MatchScore::Win->value : MatchScore::Loss->value;
            } 
            break;
        default:                                    // Must be scissors       
            // Add score for playing symbol
            $score += SymbolScore::Scissors->value;
            // Add score for win, draw or loss
            if ($elf == Symbol::Scissors) {
                $score += MatchScore::Draw->value; 
            } else {
                $score += ($elf == Symbol::Paper) ? MatchScore::Win->value : MatchScore::Loss->value;
            } 
            break;
    }
    return $score;
}

/**
 * forceResult : Choose a player symbol to force a result against a known elf symbol
 * @param Symbol $elf       Rock, Paper or Scissors
 * @param string $result    X = Lose, Y = Draw, Z = Win
 * 
 * @return Symbol           The player symbol that will give the result.
 */
function forceResult( Symbol $elf, string $result): Symbol
{
    // Assume draw
    $player = $elf;
    switch( $result ) {
        //Lose
        case 'X':
            if( $elf == Symbol::Rock) {
                $player = Symbol::Scissors;
            } else  if ( $elf == Symbol::Paper ){
                $player = Symbol::Rock;
            } else {
                $player = Symbol::Paper;
            }
            break;
        //Win
        case 'Z':
            if( $elf == Symbol::Rock) {
                $player = Symbol::Paper;
            } else  if ( $elf == Symbol::Paper ){
                $player = Symbol::Scissors;
            } else {
                $player = Symbol::Rock;
            }
            break;
        // draw
        default:
            $player = $elf;
            break;
    }
    return $player;
}

/**
 * getSymbols : Translates a line into an Elf and Player symbol.
 * 
 * if $mode == true then
 * Elf: A = Rock, B = Paper, C = Scissors
 * Player : X = Rock, Y = Paper, Z = Scissors
 * 
 * if $mode == false
 * Elf: A = Rock, B = Paper, C = Scissors
 * Result : X = Loss, Y = Draw, Z = Win so the player character is chosen to get that result given the elf character.
 *   
 * @param string $line      A string containing 2 characters (as above) separated by a space
 * 
 * @return array  Element[0] contains the Elf symbol, Element[1] contains the player symbol
 */
function getSymbols($line, $mode): array
{
    $symbols = array();

    // characters will become a 2 element array holding the 2 player characters
    $characters = explode(' ', trim($line));

    // Convert the elf character to an enum
    switch ($characters[0]) {
        case 'A':
            $symbols[0] = Symbol::Rock;
            break;
        case 'B':
            $symbols[0] = Symbol::Paper;
            break;
        case 'C':
            $symbols[0] = Symbol::Scissors;
            break;
        default:
            die( "Unexpected elf character: " . $characters[0]);
    }

    if( $mode ) {
        // Convert the player character to an enum
        switch ($characters[1]) {
            case 'X':
                $symbols[1] = Symbol::Rock;
                break;
            case 'Y':
                $symbols[1] = Symbol::Paper;
                break;
            case 'Z':
                $symbols[1] = Symbol::Scissors;
                break;
            default:
                die( "Unexpected player character: " . $characters[1]);
        }
    } else {
        $symbols[1] = forceResult( $symbols[0], $characters[1]);
    }

    return $symbols;
}



function showRound( Symbol $elf, Symbol $player, $score): void
{
    switch($elf) {
        case Symbol::Rock:
            echo "Rock";
            break;
        case Symbol::Paper:
            echo "Paper";
            break;
        case Symbol::Scissors:
            echo "Scissors";
            break;
        default:
            echo "???";
    }
    echo " V ";
    switch($player) {
        case Symbol::Rock:
            echo "Rock";
            break;
        case Symbol::Paper:
            echo "Paper";
            break;
        case Symbol::Scissors:
            echo "Scissors";
            break;
        default:
            echo "???";
    }
    echo " Score: " . $score . "\n";
}

$file = new SplFileObject("rps.txt", "r");
$totalPart1 = 0;
$totalPart2 = 0;
while( !$file->eof()) {
    $line = trim($file->fgets());
    if( strlen($line) > 0) {
        $symbols = getSymbols($line, true);
        $score = scoreRound($symbols[0], $symbols[1]);
        $totalPart1 += $score;

        $symbols = getSymbols($line, false);
        $score = scoreRound($symbols[0], $symbols[1]);
        showRound($symbols[0], $symbols[1], $score);
        $totalPart2 += $score;
    }
}
echo "Part 1\n";
echo "Total Score: " . $totalPart1 . "\n";
echo "\nPart 2\n";
echo "Total Score: " . $totalPart2 . "\n";
?>

File: day2.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <map>


using namespace std;

enum class SymbolType { Rock, Paper, Scissors };

/**
 * Base class for symbols
*/
class Symbol
{
    /**
     * Virtual functions that the concrete symbols override
    */
    protected:
    // Type ID
    virtual SymbolType getType() = 0;
    // The score for playing the symbol
    virtual int playScore() = 0;
    public:
    // Which symbol this symbol beats
    virtual SymbolType beats() = 0;
    // Which symbol this symbol draws with
    virtual SymbolType drawsWith() = 0;
    // Which symbol this symbol loses to
    virtual SymbolType losesTo() = 0;

    /**
     * Play a round against an opponent
    */
    int scoreRound( Symbol *opponent ){
        int score = playScore();
        if( *this > *opponent ) {
            score += 6;
        } else if( *this == *opponent ){
            score += 3;
        }
        return score;
    }

    /**
     * Draw
    */
    bool operator==(Symbol &rhs) {
        return( drawsWith() == rhs.getType() );
    }

    /**
     * This symbol wins 
    */
    bool operator>(Symbol &rhs) {
        return( beats() == rhs.getType() );
    }

    /**
     * This symbol loses
    */
    bool operator<(Symbol &rhs) {
        return( losesTo() == rhs.getType() );
    }
};

/**
 * Rock Symbol
*/
class Rock : public Symbol
{
    protected:
    SymbolType getType() { return SymbolType::Rock; }
    int playScore() { return 1;}
    public:
    SymbolType beats() { return SymbolType::Scissors; }
    SymbolType drawsWith() { return SymbolType::Rock; }
    SymbolType losesTo() { return SymbolType::Paper; }
};

/**
 * Paper Symbol
*/
class Paper : public Symbol
{
    protected:
    SymbolType getType() { return SymbolType::Paper; }
    int playScore() { return 2;}
    public:
    SymbolType beats() { return SymbolType::Rock; }
    SymbolType drawsWith() { return SymbolType::Paper; }
    SymbolType losesTo() { return SymbolType::Scissors; }
};

/**
 * Scissors symbol
*/
class Scissors : public Symbol
{
    protected:
    SymbolType getType() { return SymbolType::Scissors; }
    int playScore() { return 3;}
    public:
    SymbolType beats() { return SymbolType::Paper; }
    SymbolType drawsWith() { return SymbolType::Scissors; }
    SymbolType losesTo() { return SymbolType::Rock; }
};

/**
 * Create symbol classes
*/
class SymbolFactory
{
    private:
    map<int, Symbol *> symbols;
    
    public:
    /**
     * Build the symbol lookup table
    */
    SymbolFactory() {
        symbols[static_cast<int>(SymbolType::Rock)] = new Rock();
        symbols['A'] = new Rock();
        symbols['X'] = new Rock();
        symbols[static_cast<int>(SymbolType::Paper)] = new Paper();
        symbols['B'] = new Paper();
        symbols['Y'] = new Paper();
        symbols[static_cast<int>(SymbolType::Scissors)] = new Scissors();
        symbols['C'] = new Scissors();
        symbols['Z'] = new Scissors();
    }

    /**
     * Delete all the objects in the lookup table
    */
    ~SymbolFactory() {
         map<int, Symbol *>::iterator it = symbols.begin();
         while( it != symbols.end()) {
            delete it->second;
            it++;
         }
    }

    /**
     * Turn a player/elf character into a symbol
    */
    Symbol *getSymbol( int symChar) {
        return symbols[symChar]; 
    }

    /**
     * Turn a symbol type into a symbol
    */
    Symbol *getSymbol( SymbolType type ) {
        return symbols[static_cast<int>(type)];
    }

    /**
     * Get the player symbol that gives a result
    */
    Symbol *forceSymbol( Symbol *opponent, int symChar ) {
        switch( symChar ){
            // player loses
            case 'X':
                return getSymbol(opponent->beats());
            // player draws
            case 'Y':
                return getSymbol(opponent->drawsWith());
            // player wins
            case 'Z':
                return getSymbol(opponent->losesTo());
            default:
                cout << "Unexpected result character: \'" << (char)symChar << "\'";
                throw invalid_argument("Unexpected Character");
        }
    }
};


/**
 * scoreRound : Determine the player score for a round. 
 */
int scoreRound( char elfChar, char plrChar, bool mode, SymbolFactory *factory )
{
    // The symbol pointers are managed by the factory so there is no need to delete them
    Symbol *elf = factory->getSymbol((int)elfChar);
    if(mode){
        return factory->getSymbol((int)plrChar)->scoreRound(elf);
    } else {
        return factory->forceSymbol(elf, (int)plrChar)->scoreRound(elf);
    }
}

/**
 * Let's do this
*/
int main() {
    int totalPart1 = 0;
    int totalPart2 = 0;

    ifstream rps("rps.txt");
    SymbolFactory *factory = new SymbolFactory();
    string line;
    while( getline(rps, line)) {
        if( line.length() > 1) {
            totalPart1 += scoreRound(line[0], line[2], true, factory);
            totalPart2 += scoreRound(line[0], line[2], false, factory);
        }
    }
    rps.close();

    cout << "Part 1: Total Score: " << totalPart1 << endl;
    cout << "Part 2: Total Score: " << totalPart2 << endl;

    delete factory;
}